demystifying_regular_expressions/demystifying_regular_expressions.md

```diff
@@ -0,0 +1,305 @@
+
+However, in this module, we're only going to lightly cover regular expressions without your having to understand exactly how they work or how to write them.
+
+If you're interested in learning more, there are additional regular expressions modules that will teach you how to actually write and use these complex-looking codes, and we think that you may find it easier to learn than you might predict at first glance.
+
+</div>
+
+Using the pattern above, you could search for and find valid URLs (assuming we got the rules right) in your data. 
+
+In this module, we're not going to go into the rules for making a regular expression, and we won't be exhaustive in explaining each part of our complicated code. For now, here's a sneak preview of what each segment of this regular expression mean.  Don't worry about translating these or understanding them!
+
+* `http` : The literal characters "http"
+* `s?`: Optionally, the letter "s"
+* `:\/\/` : A colon and two forward slashes: "://"
+* `([A-Za-z0-9\-]+\.)?`: Optionally, a group of characters consisting of: 
+
+  - `[A-Za-z0-9\-]+` : One or more characters that's either in the A-Z range, the a-z range, the 0-9 range, or a literal hyphen (`-`)
+  - `\.` : A literal period (`.`)
+* `[A-Za-z0-9\-]+`: One or more characters that's either in the A-Z range, the a-z range, the 0-9 range, or a literal hyphen (`-`)
+* `\.` : A literal period (`.`)
+* `[A-Za-z0-9\-\.]+`: One or more characters that's either in the A-Z range, the a-z range, the 0-9 range, a literal hyphen (`-`), or a literal period (`.`)
+* `\/?`: Optionally, a final forward slash
+
+
+## A Great Tool to Start With
+
+As a final step in demystifying regular expressions, it can be helpful to introduce the online regex checker as a great resource to learn and to check your regex before trying to learn how to make your own regular expressions or use regex in SQL, R, or Python code you're writing.
+
+We like "Regex 101", at https://www.regex101.com, but you can also search online for "regex checker" to find other, similar websites.  Regex 101 allows you to enter your regular expression as well as some sample text that you want to apply the regular expression to.  
+
+Below, or by [trying it out yourself](https://regex101.com/r/QUok4D), you can see that we put in our "valid URL" regular expression (at the top of the screen, in the appropriately labeled box) as well as the different varieties of URLs, some of which we think are valid and match our pattern, and some of which we think are invalid and should fail to match our pattern.
+
+If our test strings match the pattern, they are highlighted, and if they do not match, they are not highlighted.
+
+![Regex 101 application with sample URL regex pattern and test URL strings.](media/url_regex.png)<!-- style = "border: 1px solid rgb(var(--color-highlight));" -->
```

Comment:
Weird behavior in safari browser: it allowed me to click on some of the links in the image to navigate to the websites. This led me to investigate whether this was an iframe or a normal image. This does not occur in Chrome so I don't think it is a LiaScript thing, just some weird Safari "feature" that feels kind of like a bug to me in this setting.

I don't have a suggestion for changing anything, I just wanted to record this behavior somewhere.

Reply:
WHOA!  that's weird and maybe good?  IDK